File: ios_app/iphone/ActivitiesTimerSheet.swift
import SwiftUI
import shared

extension NativeSheet {

    func showActivitiesTimerSheet(
            timerContext: ActivityTimerSheetVm.TimerContext?,
            withMenu: Bool,
            onStart: @escaping () -> Void
    ) {
        self.show { isPresented in
            ActivitiesTimerSheet(
                    isPresented: isPresented,
                    timerContext: timerContext,
                    withMenu: withMenu
            ) {
                isPresented.wrappedValue = false
                onStart()
            }
        }
    }
}

private let bgColor = c.sheetBg
private let listItemHeight = 46.0
private let topContentPadding = 4.0

private let activityItemEmojiWidth = 30.0
private let activityItemEmojiHPadding = 8.0
private let activityItemPaddingStart = activityItemEmojiWidth + (activityItemEmojiHPadding * 2)

private let secondaryFontSize = 16.0
private let secondaryFontWeight: Font.Weight = .light
private let timerHintHPadding = 5.0
private let listEngPadding = 8.0

private let myButtonStyle = MyButtonStyle()

private struct ActivitiesTimerSheet: View {
    
    @EnvironmentObject private var nativeSheet: NativeSheet
    
    @State private var vm: ActivitiesTimerSheetVm
    
    @Binding private var isPresented: Bool
    
    @State private var sheetHeight: Double
    
    @State private var isHistoryPresented = false
    @State private var isEditActivitiesPresented = false
    
    private let timerContext: ActivityTimerSheetVm.TimerContext?
    private let withMenu: Bool
    private let onStart: () -> Void
    
    init(
        isPresented: Binding<Bool>,
        timerContext: ActivityTimerSheetVm.TimerContext?,
        withMenu: Bool,
        onStart: @escaping () -> Void
    ) {
        _isPresented = isPresented
        self.timerContext = timerContext
        self.withMenu = withMenu
        self.onStart = onStart
        
        let vm = ActivitiesTimerSheetVm(timerContext: timerContext)
        _vm = State(initialValue: vm)
        
        let vmState = vm.state.value as! ActivitiesTimerSheetVm.State
        _sheetHeight = State(initialValue: calcSheetHeight(
            activitiesCount: vmState.allActivities.count,
            withMenu: withMenu
        ))
    }
    
    var body: some View {
        
        VMView(vm: vm) { state in
            
            ZStack {
                
                ScrollView {
                    
                    VStack {
                        
                        Padding(vertical: topContentPadding)
                        
                        ForEach(state.allActivities, id: \.activity.id) { activityUI in
                            
                            Button(
                                action: {
                                    // onTapGesture() / onLongPressGesture()
                                },
                                label: {
                                    
                                    ZStack(alignment: .bottomLeading) { // divider + isActive
                                        
                                        HStack {
                                            
                                            Text(activityUI.activity.emoji)
                                                .frame(width: activityItemEmojiWidth)
                                                .padding(.horizontal, activityItemEmojiHPadding)
                                                .font(.system(size: 22))
                                            
                                            Text(activityUI.listText)
                                                .foregroundColor(.primary)
                                                .truncationMode(.tail)
                                                .lineLimit(1)
                                            
                                            Spacer()
                                            
                                            TimerHintsView(
                                                timerHintsUI: activityUI.timerHints,
                                                hintHPadding: timerHintHPadding,
                                                fontSize: secondaryFontSize,
                                                fontWeight: secondaryFontWeight,
                                                onStart: {
                                                    onStart()
                                                }
                                            )
                                        }
                                        .padding(.trailing, listEngPadding)
                                        .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .leading)
                                        
                                        if state.allActivities.last != activityUI {
                                            SheetDividerBg()
                                                .padding(.leading, activityItemPaddingStart)
                                        }
                                        
                                        if activityUI.isActive {
                                            ZStack {}
                                                .frame(width: 8, height: listItemHeight - 2)
                                                .background(roundedShape.fill(c.blue))
                                                .offset(x: -4, y: -1)
                                        }
                                    }
                                    /// Ordering is important
                                    .contentShape(Rectangle()) // TRICK for tap gesture
                                    .onTapGesture {
                                        nativeSheet.showActivityTimerSheet(
                                            activity: activityUI.activity,
                                            timerContext: timerContext,
                                            hideOnStart: false,
                                            onStart: {
                                                isPresented = false
                                            }
                                        )
                                    }
                                    .onLongPressGesture(minimumDuration: 0.1) {
                                        nativeSheet.show { isActivityFormPresented in
                                            ActivityFormSheet(
                                                isPresented: isActivityFormPresented,
                                                activity: activityUI.activity
                                            ) {}
                                        }
                                    }
                                    //////
                                    .frame(alignment: .bottom)
                                }
                            )
                        }
                        .buttonStyle(myButtonStyle)
                        
                        if withMenu {
                            
                            HStack {
                                
                                ChartHistoryButton(text: "Summary", iconName: "chart.pie", iconSize: 18) {
                                    nativeSheet.show { isPresented in
                                        SummarySheet(isPresented: isPresented)
                                    }
                                }
                                .padding(.leading, 13)
                                .padding(.trailing, 12)
                                
                                ChartHistoryButton(text: "History", iconName: "list.bullet.rectangle", iconSize: 18) {
                                    isHistoryPresented = true
                                }
                                .sheetEnv(isPresented: $isHistoryPresented) {
                                    ZStack {
                                        c.bg.edgesIgnoringSafeArea(.all)
                                        HistoryView(isHistoryPresented: $isHistoryPresented)
                                    }
                                    // todo
                                    .interactiveDismissDisabled()
                                }
                                
                                Spacer()
                                
                                Button(
                                    action: { isEditActivitiesPresented = true },
                                    label: {
                                        Text("Edit")
                                            .font(.system(size: secondaryFontSize, weight: secondaryFontWeight))
                                            .padding(.trailing, timerHintHPadding)
                                    }
                                )
                                .padding(.trailing, listEngPadding)
                                .sheetEnv(
                                    isPresented: $isEditActivitiesPresented
                                ) {
                                    EditActivitiesSheet(
                                        isPresented: $isEditActivitiesPresented
                                    )
                                }
                            }
                            .frame(height: listItemHeight)
                        }
                    }
                }
            }
            .onChange(of: state.allActivities.count) { newCount in
                sheetHeight = calcSheetHeight(
                    activitiesCount: newCount,
                    withMenu: withMenu
                )
            }
        }
        .frame(maxHeight: sheetHeight)
        .background(bgColor)
        .listStyle(.plain)
        .listSectionSeparatorTint(.clear)
        .presentationDetents([.height(sheetHeight)])
        .presentationDragIndicator(.visible)
    }
}

private func calcSheetHeight(
        activitiesCount: Int,
        withMenu: Bool
) -> Double {
    // Do not be afraid of too much height because the native sheet will cut
    (listItemHeight * activitiesCount.toDouble()) +
    (withMenu ? listItemHeight : 0.0) + // Buttons
    topContentPadding
}

private struct MyButtonStyle: ButtonStyle {

    func makeBody(configuration: Self.Configuration) -> some View {
        configuration
                .label
                .frame(height: listItemHeight)
                .background(configuration.isPressed ? c.dividerFg : bgColor)
    }
}

private struct ChartHistoryButton: View {

    let text: String
    let iconName: String
    let iconSize: CGFloat
    let onClick: () -> Void

    var body: some View {
        Button(
                action: { onClick() },
                label: {
                    HStack {
                        Image(systemName: iconName)
                                .font(.system(size: iconSize, weight: .thin))
                                .padding(.trailing, 3 + onePx)
                        Text(text)
                                .font(.system(size: secondaryFontSize, weight: secondaryFontWeight))
                    }
                }
        )
    }
}

File: android_app/src/main/java/me/timeto/app/ui/ActivitiesTimerSheet.kt
package me.timeto.app.ui

import androidx.annotation.DrawableRes
import androidx.compose.foundation.ExperimentalFoundationApi
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.combinedClickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material.Icon
import androidx.compose.material.Text
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.platform.LocalConfiguration
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import me.timeto.app.*
import me.timeto.app.R
import me.timeto.shared.vm.ActivityTimerSheetVM
import me.timeto.shared.vm.ActivitiesTimerSheetVM

fun ActivitiesTimerSheet__show(
    timerContext: ActivityTimerSheetVM.TimerContext?,
    withMenu: Boolean,
) {
    Sheet.show { layer ->
        ActivitiesTimerSheet(
            layerActivitiesSheet = layer,
            timerContext = timerContext,
            withMenu = withMenu,
        )
    }
}

private val activityItemEmojiHPadding = 8.dp
private val activityItemEmojiWidth = 32.dp
private val activityItemPaddingStart = activityItemEmojiWidth + (activityItemEmojiHPadding * 2)

private val listItemHeight = 42.dp
private val topContentPadding = 2.dp
private val bottomContentPadding = 4.dp

private val secondaryFontSize = 14.sp
private val secondaryFontWeight = FontWeight.Light
private val timerHintHPadding = 5.dp
private val listEngPadding = 8.dp

@OptIn(ExperimentalFoundationApi::class)
@Composable
private fun ActivitiesTimerSheet(
    layerActivitiesSheet: WrapperView.Layer,
    timerContext: ActivityTimerSheetVM.TimerContext?,
    withMenu: Boolean,
) {

    val (_, state) = rememberVm(timerContext) { ActivitiesTimerSheetVM(timerContext) }

    val screenHeight = LocalConfiguration.current.screenHeightDp.dp
    val contentHeight = (listItemHeight * state.allActivities.size) +
                        (if (withMenu) listItemHeight else 0.dp) + // Buttons
                        topContentPadding +
                        bottomContentPadding

    LazyColumn(
        modifier = Modifier
            .background(c.sheetBg)
            .navigationBarsPadding()
            .height(contentHeight.limitMax(screenHeight - 60.dp))
            .fillMaxWidth(),
        contentPadding = PaddingValues(top = topContentPadding, bottom = bottomContentPadding)
    ) {

        items(state.allActivities) { activityUI ->

            val activity = activityUI.activity

            Box(
                contentAlignment = Alignment.BottomCenter, // for divider
            ) {

                Row(
                    modifier = Modifier
                        .height(listItemHeight)
                        .combinedClickable(
                            onClick = {
                                ActivityTimerSheet__show(
                                    activity = activity,
                                    timerContext = timerContext,
                                ) {
                                    layerActivitiesSheet.close()
                                }
                            },
                            onLongClick = {
                                Sheet.show { layer ->
                                    ActivityFormSheet(
                                        layer = layer,
                                        activity = activityUI.activity,
                                    )
                                }
                            },
                        )
                        .padding(end = listEngPadding),
                    verticalAlignment = Alignment.CenterVertically,
                ) {

                    Text(
                        text = activity.emoji,
                        modifier = Modifier
                            .padding(horizontal = activityItemEmojiHPadding)
                            .width(activityItemEmojiWidth),
                        textAlign = TextAlign.Center,
                        fontSize = 20.sp,
                    )

                    Text(
                        activityUI.listText,
                        modifier = Modifier
                            .weight(1f),
                        color = c.text,
                        textAlign = TextAlign.Start,
                        overflow = TextOverflow.Ellipsis,
                        maxLines = 1,
                    )

                    TimerHintsView(
                        modifier = Modifier,
                        timerHintsUI = activityUI.timerHints,
                        hintHPadding = timerHintHPadding,
                        fontSize = secondaryFontSize,
                        fontWeight = secondaryFontWeight,
                        onStart = {
                            layerActivitiesSheet.close()
                        }
                    )
                }

                SheetDividerBg(
                    modifier = Modifier
                        .padding(start = activityItemPaddingStart),
                    isVisible = state.allActivities.lastOrNull() != activityUI,
                )

                if (activityUI.isActive)
                    Box(
                        modifier = Modifier
                            .align(Alignment.CenterStart)
                            .offset(x = (-4).dp)
                            .height(listItemHeight - 2.dp)
                            .clip(roundedShape)
                            .background(c.blue)
                            .width(8.dp)
                    ) {}
            }
        }

        if (withMenu) {

            item {

                HStack(
                    modifier = Modifier
                        .height(listItemHeight)
                        .padding(start = 8.dp),
                    verticalAlignment = Alignment.CenterVertically,
                ) {

                    ChartHistoryButton(
                        text = "Summary",
                        iconResId = R.drawable.sf_chart_pie_small_thin,
                        iconSize = 17.dp,
                    ) {
                        Sheet.show(
                            topPadding = 4.dp,
                        ) { layer ->
                            SummarySheet(layer)
                        }
                    }

                    ChartHistoryButton(
                        "History",
                        iconResId = R.drawable.sf_list_bullet_rectangle_small_thin,
                        iconSize = 19.dp,
                        extraIconPadding = onePx,
                    ) {
                        Dialog.show(
                            modifier = Modifier.fillMaxHeight(0.95f),
                        ) { layer ->
                            HistoryDialogView(layer::close)
                        }
                    }

                    SpacerW1()

                    Text(
                        text = "Edit",
                        modifier = Modifier
                            .padding(end = listEngPadding)
                            .clip(squircleShape)
                            .clickable {
                                Sheet.show { layer ->
                                    EditActivitiesSheet(layer = layer)
                                }
                            }
                            .padding(horizontal = timerHintHPadding, vertical = 4.dp),
                        color = c.blue,
                        fontSize = secondaryFontSize,
                        fontWeight = secondaryFontWeight,
                    )
                }
            }
        }
    }
}

@Composable
private fun ChartHistoryButton(
    text: String,
    @DrawableRes iconResId: Int,
    iconSize: Dp,
    extraIconPadding: Dp = 0.dp,
    onClick: () -> Unit,
) {
    HStack(
        modifier = Modifier
            .clip(squircleShape)
            .clickable {
                onClick()
            }
            .padding(horizontal = 8.dp, vertical = 4.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {

        Icon(
            painterResource(iconResId),
            contentDescription = text,
            tint = c.blue,
            modifier = Modifier
                .padding(end = 5.dp + extraIconPadding)
                .size(iconSize)
        )

        Text(
            text = text,
            color = c.blue,
            fontSize = secondaryFontSize,
            fontWeight = secondaryFontWeight,
        )
    }
}
